Abc

cdonovick opened the pull request:
Adds `bit_vector_abc.py` with classes `AbstractBitVector` and `AbstractBitVectorMeta`

Forces BitVector types to be bound to a specific size using the syntax `BitVectorT[size]`

Construction of BitVectors from `int` with implicit size  is still supported

Uses "abstractness" of the method `_size` to determine if a BitVector type is bound to a size.  It might be more appropriate to add some other attribute or inherit from some marker class instead.

cdonovick committed:
bit_vector_abc

cdonovick committed:
Merge branch 'master' into abc

cdonovick committed:
Update abc

cdonovick committed:
Rename BitVecAbc; remove seitem from base ABC

cdonovick committed:
Move to BitVector[size] syntax

cdonovick commented:
@leonardt @phanrahan

cdonovick committed:
Remove format string

cdonovick committed:
Fix some bugs / add some tests

cdonovick committed:
Fix broken ite test

cdonovick committed:
Fix how bases are generated in meta.__getitem__; add tests

cdonovick committed:
Remove extraneous if

cdonovick commented:
Should be noted `AbstractBitVectorMeta` is really `AbstractFixedSizedMeta` and really doesn't do anything but construct classes with the a method called `_size`

cdonovick committed:
add support for inheriting from sized types

leonardt commented on bit_vector/bit_vector.py:
```diff
@@ -159,64 +158,55 @@ def __setitem__(self, index, value):
+            return BitVector[len(v)](v)
         else:
-            return self.bits()[index]
+            return BitVector[1](self.bits()[index])
```
Should this be a `Bit` object instead? (We will have to define this)

cdonovick replied:
I will add a bit type

leonardt commented on bit_vector/bit_vector.py:
```diff
@@ -159,64 +158,55 @@ def __setitem__(self, index, value):
     @binary
     def bvashr(self, other):
-        return BitVector( self.as_sint() >> other.as_uint(), num_bits=self.num_bits)
+        return type(self)( self.as_sint() >> other.as_uint())
```
We could clean this up by moving all casting logic to a decorator (or add it to the existing decorator), but I don't think this is a blocking issue.

cdonovick replied:
I dislike this because it makes the behavior kind of opaque in my mind.

leonardt commented on bit_vector/bit_vector.py:
```diff
@@ -239,54 +229,24 @@ def bvcomp(self, other):
-    @binary
-    def bvsge(self, other):
-        return BitVector(self.as_sint() >= other.as_sint(), num_bits=1)
+        return BitVector[1](self.as_sint() < other.as_sint())
```
Did we mean to drop all these comparisons?

leonardt replied:
Nevermind, these are defined in the abstract class in terms of other methods which have been implemented

leonardt commented on bit_vector/bit_vector.py:
```diff
@@ -296,84 +256,74 @@ def adc(a, b, c):
         if other == 0:
-            return BitVector(1, num_bits=self.num_bits)
-        return BitVector(self.as_uint() // other, num_bits=self.num_bits)
+            return type(self)((1 << self.size) - 1)
```
Is this the expected semantics for div by zero? I remember you mentioned it was a special case, just wanted to double check.

cdonovick replied:
This is the SMT-lib semantics

leonardt commented on bit_vector/bit_vector.py:
```diff
@@ -440,7 +387,8 @@ def sext(self, other):
         """
-        return self.concat(BitVector(other.as_uint() * [self[-1]]), self)
+        ext = other.as_uint()
+        return self.concat(BitVector[ext](ext * [self[-1]]), self)
```
I think this is changing the interface to `sext`. Before, `other` indicated how many bits to extend the current bit vector. Now it uses `ext` as the size of the target bit vector. We should decide on a semantics since if we want to change it to be this, this is a good time to do it (also is this the SMT semantics?).

cdonovick replied:
The SMT-lib zero and sign extend have the semantics we currently have.

leonardt reviewed (commented):
Some comments, still need to read through the abstract class definition

cdonovick committed:
Add AbstractBit, type annotations

cdonovick committed:
Rewite bitvector

Change signature of most operators from
OP :: BitVector[k] -> BitVector[j] -> BitVector[k]
to
OP :: BitVector[k] -> BitVector[k] -> BitVector[k]

cdonovick committed:
Use type(self) instead of explicit types

cdonovick committed:
Add support for old style bitvectors

cdonovick committed:
Bug fix

cdonovick committed:
Add TypeFamily

cdonovick reviewed (commented):
I am not in love with how I am currently supporting the type family concept but it gets the job done.

leonardt commented on bit_vector/bit_vector_abc.py:
```diff
@@ -0,0 +1,299 @@
+        if not isinstance(idx, int):
+            raise TypeError()
+        if idx < 0:
+            raise ValueError('Size of BitVectors must be positive')
```
Did we decide to allow size 0?

cdonovick replied:
I think we decided against it but forbidding them broke some tests and create weird corner cases for user code.  I think its better to allow them.

leonardt replied:
Okay, let's leave it for now but keep on eye on whether this becomes a problem.

leonardt reviewed (approved):
LGTM, finally had a chance to read the ABC implementation, makes enough sense that I think could help debug any issues that come up. Let ship this!
